### PR TITLE
osd: fix hangs during shutdown

### PR DESCRIPTION
**Pull request by dzafman**
- [X] References tracker ticket
-  ~~Updates documentation if necessary~~
-  ~~Includes tests for new functionality or reproducer for bug~~

### CODE REVIEW

**Comment by tchaikov** [comment]
cc @majianpeng

**Comment by xiexingguo** [comment]
Ahh, I run into a similar issue a couple of days ago (though need further investigation):


> I also took a close look at the failure case at: http://qa-proxy.ceph.com/teuthology/xxg-2018-09-21_02:02:01-rados-wip-fix-polog-overtrim-distro-basic-smithi/3047956/, which revealed that bluestore were starting dropping callbacks(e.g., on_commit) from high-level callers?

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -10225,23 +10225,26 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)
...
+      sdata->shard_lock.Unlock();
+      return;    // OSD shutdown, discard.
+    }
     sdata->shard_lock.Unlock();
     handle_oncommits(oncommits);
     return;
   }
 
   OpQueueItem item = sdata->pqueue->dequeue();
+  if (osd->is_stopping()) {
```
Comment (majianpeng) [inline src/osd/OSD.cc]: If there are multi item in pqueue, this code will do multi times. Is if to clean pqueue and context_queue to avoid this.

Comment (dzafman) [reply src/osd/OSD.cc]: @majianpeng I thought it was safer to just return the code to working exacting the way it did before it was broken.

**Pull request merged**